project/src/main/java/com/google/sps/data/VisionLibrary.java

```diff
@@ -0,0 +1,330 @@
+package com.google.sps.data;
+import com.google.gson.Gson;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Date;
+import java.awt.Color;
+import java.util.stream.Collectors;  
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+
+import com.google.cloud.vision.v1.*;
+import com.google.cloud.vision.v1.WebDetection.WebEntity;
+import com.google.cloud.vision.v1.WebDetection.WebLabel;
+import com.google.cloud.vision.v1.WebDetection.WebPage;
+import com.google.protobuf.ByteString;
+import java.io.ByteArrayOutputStream;
+
+import java.util.Random;
+import java.math.BigDecimal; 
+
+
+public class VisionLibrary {
+  public static AnnotateImageResponse handleCloudVisionRequest(byte[] imageBytes, List<Feature> allFeatures) throws IOException {
+    // Convert image bytes into proper image format.
+    ByteString byteString = ByteString.copyFrom(imageBytes);
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    // Create a request with all the features desired.
+    AnnotateImageRequest request = AnnotateImageRequest.newBuilder().addAllFeatures(allFeatures).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    // Connect to cloud vision client and send request.
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+    client.close();
+
+    // Receive the reponse and handle potential errors.
+    List<AnnotateImageResponse> imageResponses = batchResponse.getResponsesList();
+    AnnotateImageResponse imageResponse = imageResponses.get(0);
+    if (imageResponse.hasError()) {
+      System.err.println("Error getting image labels: " + imageResponse.getError().getMessage());
+      return null;
+    }
+
+    return imageResponse;
+  }
+
+  public static String formatImageResponse(ImagesService imagesService, Gson gson, AnnotateImageResponse imageResponse, BlobKey blobKey) {
+    // These are labels for the overall image based on what the vision api thinks.
+    List<EntityAnnotation> labelAnnotations = imageResponse.getLabelAnnotationsList();
```

Comment:
You should check for null here, since it's possible imageResponse is null (see line 63 above).